---
title: "Indepth Analysis of Updated version of the ONT variant callset"
date: '`r Sys.Date()`'
output: 
    bookdown::html_document2:
        toc: true
        toc_float: true
        df_print: paged
        code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(ggrepel)
source(here("scripts", "annotation-phredticks.R"))
```

# Objective
In depth analysis of an example submission to demonstrate the utility of the stratifications. 

Will compare stratified results for Kishwar's ONT, `PEPPER-DeepVariant (ONT_9to6)` and `PEPPER-DeepVariant (ONT)` to the updated version using winnowmap for a supplemental result.

Where does the new method improve relative to the previous run

# Approach
Compared stratified benchmarking results across datasets. 

# Loading Data
```{r}
submissions <- c(pepper = "RU88N") %>% 
    rep(each = 3)

giab_ids <- paste0("HG00",2:4)


pepper_v1_ids <- map2(submissions, giab_ids, ~paste0(.x, "_",.y))
pepper_v1_names <- map2(names(submissions), giab_ids, ~paste0(.x, "_",.y))

pepper_v1_csvs <- map2(submissions, pepper_v1_ids,
                      ~here("data", "extended_csvs", 
                           .x, paste0(.y, ".extended.csv"))
                      ) %>% 
    set_names(pepper_v1_names)

pepper_v2_csvs <- map(giab_ids,
                      ~here("data-raw", "Winnowmap_pepper_deepvariant_improved",
                            paste0(., "_pFDA_ONT_winnowmap_pepper_deepvariant.vcf.gz_extended.csv"))
                      ) %>% 
    set_names(paste0("update_",giab_ids))


## Generating combined list
extended_csvs <- c(pepper_v1_csvs, pepper_v2_csvs)

## Loading extended CSVs #######################################################
## Defining col types to avoid ti/tiv logical issues
ctypes <- str_c(c(rep("c", 7), rep("d", 9), rep("dccddcd", 7)), 
                sep = "", collapse = "")

extended_df <- extended_csvs %>% 
    map_dfr(read_csv, col_types = ctypes, .id = "callset")

results_df <- extended_df %>% 
    filter(Subtype == "*", Filter == "PASS",  Subset.Size > 1000) %>% 
    mutate(phred_F1 = -10*log(1-METRIC.F1_Score)/log(10))


## Calculating geometric mean and converting metrics to phred scale
mean_metrics_df <- results_df %>% 
    separate(callset, into = c("caller","giab_id"), sep = "_") %>% 
    filter(giab_id != "HG002") %>% 
    select(giab_id, caller, Type, Subset,
           METRIC.Recall, METRIC.Precision, METRIC.F1_Score) %>% 
    pivot_longer(cols = c("METRIC.Recall","METRIC.Precision", "METRIC.F1_Score"),
                 names_to = "metric", values_to = "value") %>% 
    group_by(caller, Type, Subset, metric) %>% 
    summarise(value = exp(mean(log(value)))) %>% 
    mutate(phred_value = -10*log(1-value)/log(10))
```


Stratification Metadata
```{r}
strat_metadata_df <- read_tsv(here("data-raw","v2.0-GRCh38-strat-metadata.txt")) %>% 
    mutate(strat_group = str_extract(path, ".*(?=/)"))
```
Stratifications per group
```{r} 
strat_metadata_df %>% 
    count(strat_group)
```

```{r}
strat_metadata_df %>% 
    filter(strat_group == "LowComplexity") %>% 
    mutate(subgroup = case_when( str_detect(path, "notin") ~ "Union",
                                 str_detect(path, "AllTandemRepeatsandHomopolymers") ~ "Union",
                                 str_detect(path, "TandemRepeats") ~ "TandemRepeats",
                                 str_detect(path, "All|notin") ~ "Union",
                                 str_detect(path, "homopol") ~ "Homopolymers",
                                 str_detect(path, "SimpleRepeat") ~ "SimpleRepeats",
                                TRUE ~ "Unassigned"
                                ))  %>% 
    count(subgroup)
```


Making a 2D scatter to compare the variant calling methods
```{r}
wide_callset_results_df <- results_df %>% 
    filter(str_detect(Subset, "HG00", negate = TRUE),
           Subset != "TS_boundary") %>% 
    select(callset, Type, Subset, phred_F1) %>% 
    separate(callset, into = c("caller","giab_id"), sep = "_") %>% 
    pivot_wider(names_from = "caller", values_from = "phred_F1") %>% 
    left_join(strat_metadata_df)
```

Annotating with metadata
```{r}
strat_results_df <- mean_metrics_df %>%
    left_join(strat_metadata_df) %>% 
    filter(str_detect(Subset, "HG00", negate = TRUE),
           Subset != "TS_boundary", Subset != "TS_Contained")

wide_strat_results_df <- strat_results_df %>% 
    select(-value) %>% 
    pivot_wider(names_from = "caller", values_from = "phred_value")
    
```

```{r}
wide_strat_results_df %>% 
    count(strat_group)
```

# Results 
```{r}
wide_strat_results_df %>%
    select(strat_id, strat_group) %>%
    distinct() %>%
    arrange(strat_group)
```

```{r}
wide_strat_results_df
```

```{r}
scatter_plt_df <- wide_strat_results_df %>%
        ## Reducing the number of plotted stratifications
        filter(strat_group != "GCcontent",
               strat_group != "FunctionalTechnicallyDifficultRegions",
               strat_group != "union",
               str_detect(strat_id, "di|tri|quadTR", negate = TRUE),
               str_detect(strat_id, "All|notin|non", negate = TRUE)) %>% 
    mutate(metric_diff = pepper - update) %>% 
    filter(pepper != 0, update != 0) %>% 
    group_by(Type, metric) %>% 
    arrange(-metric_diff) %>% 
    mutate(rank = 1:n(), 
           labeled_strats = case_when(rank < 3 && metric_diff > 0 ~ strat_id, 
                                      n() - rank < 3 ~ strat_id)) %>% 
    mutate(metric = str_remove(metric, "METRIC."),
           metric = if_else(metric == "F1_Score", "F1", metric)) %>% 
    mutate(Type = if_else(Type == "SNP", "SNV", Type))
    
(
    scatter_plt <- scatter_plt_df %>%
        ggplot() +
        geom_abline(aes(intercept = 0, slope = 1)) +
        geom_point(
            aes(
                x = pepper,
                y = update,
                fill = strat_group,
                label = strat_id
            ),
            shape = 21
        ) +
        geom_text_repel(
            point.padding = 0.25,
            aes(x = pepper,
                y = update,
                label = labeled_strats),
            size = 2,
            na.rm = TRUE,
            min.segment.length = 0.2,
            max.overlaps = 50
        ) +
        facet_grid(metric ~ Type) +
        theme_bw() +
        labs(x = "Challenge Submission %",
             y = "Updated Version %",
             fill = "Stratification\nCategory") +
        scale_x_continuous(
            breaks = c(0, 10, 20, 30, 40),
            labels = c("0", "90", "99", "99.9", "99.99")
        ) +
        scale_y_continuous(
            breaks = c(0, 10, 20, 30, 40),
            labels = c("0", "90", "99", "99.9", "99.99")
        ) +
        annotation_phredticks(sides = "lb", outside = TRUE,
                          scaled = FALSE) +
        scale_fill_brewer(type = "qual", palette = 3) +
        coord_equal(clip = "off") +
        theme(
            strip.text.y = element_text(angle = 360),
            legend.position = "bottom"
        ) +
        guides(fill = guide_legend(
            title.position = "left", nrow = 2
        ))
)
```

```{r}
ggsave(here("figures","indepth_update.png"), scatter_plt, height = 8, width = 6, dpi = "retina")
```
```{r}
plotly::ggplotly(scatter_plt)
```





# Session Information
## System Information
```{r}
sessioninfo::platform_info()
```


## Package Versions
```{r}
sessioninfo::package_info() %>% 
    filter(attached = TRUE) %>% 
    select(package, loadedversion, date, source) %>%
    knitr::kable(booktabs = TRUE, row.names = FALSE)
```